utils/buildutils.go

```
@@ -0,0 +1,423 @@
+package utils
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	buildinfo "github.com/jfrog/build-info-go/entities"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+)
+
+const BuildInfoDetails = "details"
+
+func getBuildDir(buildName, buildNumber, projectKey, buildsDirPath string) (string, error) {
+	encodedDirName := base64.StdEncoding.EncodeToString([]byte(buildName + "_" + buildNumber + "_" + projectKey))
+	buildsDir := filepath.Join(buildsDirPath, encodedDirName)
+	err := os.MkdirAll(buildsDir, 0777)
+	if err != nil {
+		return "", err
+	}
+	return buildsDir, nil
+}
+
+func getPartialsBuildDir(buildName, buildNumber, projectKey, buildsDirPath string) (string, error) {
+	buildDir, err := getBuildDir(buildName, buildNumber, projectKey, buildsDirPath)
+	if err != nil {
+		return "", err
+	}
+	buildDir = filepath.Join(buildDir, "partials")
+	err = os.MkdirAll(buildDir, 0777)
+	if err != nil {
+		return "", err
+	}
+	return buildDir, nil
+}
+
+func SaveBuildInfo(buildName, buildNumber, projectKey, buildsDirPath string, buildInfo *buildinfo.BuildInfo, log Log) error {
+	b, err := json.Marshal(buildInfo)
+	if err != nil {
+		return err
+	}
+	var content bytes.Buffer
+	err = json.Indent(&content, b, "", "  ")
+	if err != nil {
+		return err
+	}
+	dirPath, err := getBuildDir(buildName, buildNumber, projectKey, buildsDirPath)
+	if err != nil {
+		return err
+	}
+	log.Debug("Creating temp build file at: " + dirPath)
+	tempFile, err := ioutil.TempFile(dirPath, "temp")
+	if err != nil {
+		return err
+	}
+	defer tempFile.Close()
```

Let's not ignore the error received by this function.